File: matrix.py
from copy import copy
from random import shuffle

__author__ = 'wesley'


class Matrix():
    """Data Structure for 2-dimensional data.  You can load and
    save this data structure to files.  You can also slice this
    data in various ways using .submatrix()"""
    def __init__(self, rows=0, columns=0, default=0.0):
        self.elements = [[default for _ in range(columns)] for _ in range(rows)]
        self.row_labels = []
        self.column_labels = []

    def __getitem__(self, item):
        return self.elements[item]

    def __len__(self):
        return len(self.elements)

    def columns(self):
        return len(self.elements[0])

    def column(self, index):
        return [row[index] for row in self.elements]

    def save(self, filename):
        """Save matrix to tab-seperated file"""
        with open(filename, 'w') as f:
            # Write Header
            header = '\t'.join(self.column_labels) + '\n'
            f.write(header)
            # Write elements
            for i, row in enumerate(self.elements):
                string = self.row_labels[i]
                string += '\t'.join(map(float, row))
                f.write(string)

    def load(self, filename, datatype=float, col_headers=True, row_headers=True):
        """
        load a matrix from a tab-seperated file
        :param filename: string path of file to load
        :param col_headers: true if first row is column_labels
        :param row_headers: true if first element in row is the label
        :param datatype: type to cast values to, i.e. int, float, etc.
        :param col_headers: boolean. True if there are column labels on first row
        :return: None
        """
        self.elements = []
        with open(filename) as f:
            for i, line in enumerate(f):
                tokens = line.strip().split('\t')
                if len(tokens) == 0:  # Handle blank lines
                    continue
                if col_headers and i == 0:
                    self.column_labels = tokens
                else:
                    if row_headers:
                        self.row_labels.append(tokens[0])
                        tokens = tokens[1:]
                    row = [datatype(value) for value in tokens]
                    self.elements.append(row)

    def flatten(self):
        """flatten the matrix into a 1d array"""
        pass

    def dimensions(self):
        return len(self.elements), len(self.elements[0])

    def submatrix(self, rows, columns):
        """
        return a submatrix (does not modify matrix). order DOES matter
        :param rows: list of rows to be included in sub-matrix
        :param columns: list of columns to be included in sub-matrix
        :return:
        """
        s = Matrix(len(rows), len(columns))
        s.elements = [[self.elements[i][j] for j in columns] for i in rows]
        return s

    def transpose(self):
        rows, cols = self.dimensions()
        result = Matrix(cols, rows)
        result.elements = [[self.elements[i][j] for i in range(rows)] for j in range(cols)]
        return result

    def random_split(self):
        N = len(self)
        rows = list(range(len(self.elements)))
        cols = list(range(self.columns()))
        shuffle(rows)
        a = rows[:N//2]
        b = rows[N/2:]
        A = self.submatrix(a, cols)
        B = self.submatrix(b, cols)
        assert(len(A) + len(B) == N)
        return A, B

    def split(self, column, value):
        left = Matrix()
        right = Matrix()
        for row in self.elements:
            if row[column] < value:
                left.elements.append(copy(row))
            else:
                right.elements.append(copy(row))
        return left, right


def test_matrix():
    m = Matrix(4, 3, 0.0)
    m[0][2] = 1.0
    m[1][2] = 2.0
    m[2][2] = 3.0
    m[3][2] = 4.0
    print(m.elements)
    print(m.elements[3])
    print(m.column(2))
    s = m.submatrix(range(4), [2])
    print(s.elements)
    print()
    print(m.transpose().elements)

File: matrix.cpp
#include "matrix.hpp"
#include "util.hpp"

#include <cassert>
#include <cstdlib>

#include <fstream>
#include <string>
#include <iostream>
using namespace std;

Matrix::Matrix() {
	
}

void Matrix::load(string filename, bool use_column_labels, bool use_row_labels) {
	ifstream file(filename.c_str());
	string line;
	int line_number = 0;
	while(getline(file, line)) {
		vector<string> tokens = split_string(line, "\t");
		if(tokens.size() == 0) {
			cout << "Matrix.load(): skipping blank line on line #" << line_number << endl;
			line_number++;
			continue;
		}
		if(use_column_labels && line_number == 0) {
			column_labels = tokens;
			line_number++;
			continue;
		}
		else {
			if(use_row_labels) {
				row_labels.push_back(tokens[0]);
				tokens.erase(tokens.begin()+0);
			}
			vector<double> row;
			for(int i = 0; i < tokens.size(); i++) {
				double element = atof(tokens[i].c_str());
				row.push_back(element);
			}
			elements.push_back(row);
		}
		line_number++;
	}
	file.close();
}

void Matrix::save(string filename) {
	ofstream file(filename.c_str());
	//Write column header
	if(column_labels.size() > 0) {
		file << join(column_labels, '\t');
	}
	file << endl;
	//Write elements
	for(int i = 0; i < elements.size(); i++) {
		vector<double> & row = elements[i];
		if(row_labels.size() > 0) {
			file << row_labels[i] << '\t';
		}
		file << join(row, '\t');
		file << endl;
	}
	file.close();
}

int Matrix::columns() {
	if(elements.size() == 0) {
		return 0;
	}
	else {
		return elements[0].size();
	}
}

int Matrix::rows() {
	return elements.size();
}

vector<double> & Matrix::operator[](int i) {
	assert(i < elements.size());
	return elements[i];
}

vector<double> Matrix::column(int index) {
	vector<double> result;
	if(index < 0) {
		index += columns();
	}
	for(int i = 0; i < rows(); i++) {
		double element = elements[i][index];
		result.push_back(element);
	}
	return result;
}

/*
Return a subset of the matrix's elements.
rows is a list of row indices
columns is a list of column indices
i.e.
m = [[0,1],[2,3],[4,5]]
m.submatrix([0,1],[1]) = [[1],[3]]
*/
Matrix Matrix::submatrix(vector<int> rows, vector<int> columns) {
	Matrix m;
	for(int j = 0; j < columns.size(); j++) {
		m.column_labels.push_back(column_labels[j]);
	}
	for(int i = 0; i < rows.size(); i++) {
		int y = rows[i];
		vector<double> row;
		m.row_labels.push_back(row_labels[y]);
		for(int j = 0; j < columns.size(); j++) {
			int x = columns[j];
			row.push_back(elements[y][x]);
		}
		m.elements.push_back(row);
	}
	return m;
}

/*
Split the matrix into two based on a column's value
i.e.
m = [[0,1],[0,2],[0,3]]
Matrix m1, m2;
m.split(1, 2, m1, m2);
m1 = [[0,1]]
m2 = [[0,2],[0,3]]
*/
void Matrix::split(int column_index, double value, Matrix & m1, Matrix & m2) {
	vector<int> m1_rows;
	vector<int> m2_rows;
	for(int i = 0; i < elements.size(); i++) {
		double element = elements[i][column_index];
		if(element < value) {
			m1_rows.push_back(i);
		}
		else {
			m2_rows.push_back(i);
		}
	}
	vector<int> all_cols = range(columns());
	m1 = submatrix(m1_rows, all_cols);
	m2 = submatrix(m2_rows, all_cols);
}

